File: CPP/YanaPOnlineUtil/Stream/MemorySizeCaliculator.cpp
#include "MemorySizeCaliculator.h"
#include <string.h>

namespace YanaPOnlineUtil
{
namespace Stream
{

// コンストラクタ
MemorySizeCaliculator::MemorySizeCaliculator()
	: Size(0)
{
}

// intのシリアライズ
bool MemorySizeCaliculator::Serialize(int *pData)
{
	Size += sizeof(int);
	return true;
}

// unsigned intのシリアライズ
bool MemorySizeCaliculator::Serialize(unsigned int *pData)
{
	Size += sizeof(unsigned int);
	return true;
}

// shortのシリアライズ
bool MemorySizeCaliculator::Serialize(short *pData)
{
	Size += sizeof(short);
	return true;
}

// unsigned shortのシリアライズ
bool MemorySizeCaliculator::Serialize(unsigned short *pData)
{
	Size += sizeof(unsigned short);
	return true;
}

// charのシリアライズ
bool MemorySizeCaliculator::Serialize(char *pData)
{
	Size += sizeof(char);
	return true;
}

// unsigned charのシリアライズ
bool MemorySizeCaliculator::Serialize(unsigned char *pData)
{
	Size += sizeof(unsigned char);
	return true;
}

// floatのシリアライズ
bool MemorySizeCaliculator::Serialize(float *pData)
{
	Size += sizeof(float);
	return true;
}

// 文字列のシリアライズ
bool MemorySizeCaliculator::Serialize(std::string *pData)
{
	Size += sizeof(unsigned int);	// 文字列長.
	Size += pData->length();		// 文字数.
	return true;
}

}
}

File: CSharp/YanaPOnlineUtil/Stream/MemorySizeCaliculator.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace YanaPOnlineUtil.Stream
{
	/// <summary>
	/// メモリサイズ計算
	/// </summary>
	public class MemorySizeCaliculator : IMemoryStream
	{

		/// <summary>
		/// サイズ
		/// </summary>
		public int Size { get; private set; }

		/// <summary>
		/// コンストラクタ
		/// </summary>
		public MemorySizeCaliculator()
		{
			Size = 0;
		}

		/// <summary>
		/// シリアライズ
		/// </summary>
		/// <param name="Data">データ</param>
		/// <returns>成功したらtrueを返す。</returns>
		public bool Serialize(ref int Data)
		{
			Size += sizeof(int);
			return true;
		}

		/// <summary>
		/// シリアライズ
		/// </summary>
		/// <param name="Data">データ</param>
		/// <returns>成功したらtrueを返す。</returns>
		public bool Serialize(ref uint Data)
		{
			Size += sizeof(uint);
			return true;
		}

		/// <summary>
		/// シリアライズ
		/// </summary>
		/// <param name="Data">データ</param>
		/// <returns>成功したらtrueを返す。</returns>
		public bool Serialize(ref short Data)
		{
			Size += sizeof(short);
			return true;
		}

		/// <summary>
		/// シリアライズ
		/// </summary>
		/// <param name="Data">データ</param>
		/// <returns>成功したらtrueを返す。</returns>
		public bool Serialize(ref ushort Data)
		{
			Size += sizeof(ushort);
			return true;
		}

		/// <summary>
		/// シリアライズ
		/// </summary>
		/// <param name="Data">データ</param>
		/// <returns>成功したらtrueを返す。</returns>
		public bool Serialize(ref char Data)
		{
			Size += 1;
			return true;
		}

		/// <summary>
		/// シリアライズ
		/// </summary>
		/// <param name="Data">データ</param>
		/// <returns>成功したらtrueを返す。</returns>
		public bool Serialize(ref byte Data)
		{
			Size += 1;
			return true;
		}
		/// <summary>
		/// シリアライズ
		/// </summary>
		/// <param name="Data">データ</param>
		/// <returns>成功したらtrueを返す</returns>

		public bool Serialize(ref string Data)
		{
			// 文字列長
			Size += sizeof(uint);

			// 文字列
			Size += Data.Length;
			return true;
		}

		/// <summary>
		/// シリアライズ
		/// </summary>
		/// <param name="Data">データ</param>
		/// <returns>成功したらtrueを返す</returns>
		public bool Serialize(char[] Data)
		{
			Size += Data.Length;
			return true;
		}
	}
}
